phpBB/phpbb/notification/method/email.php

```diff
@@ -68,11 +78,87 @@ public function is_available(type_interface $notification_type = null)
+	{
+		$notified_users = array();
+
+		$sql = 'SELECT user_id
+			FROM ' . $this->notification_emails_table . '
+			WHERE notification_type_id = ' . (int) $notification_type_id .
+			(isset($options['item_id']) ? ' AND item_id = ' . (int) $options['item_id'] : '') .
+			(isset($options['item_parent_id']) ? ' AND item_parent_id = ' . (int) $options['item_parent_id'] : '') .
+			(isset($options['user_id']) ? ' AND user_id = ' . (int) $options['user_id'] : '');
+		$result = $this->db->sql_query($sql);
+		while ($row = $this->db->sql_fetchrow($result))
+		{
+			$notified_users[$row['user_id']] = $row;
+		}
+		$this->db->sql_freeresult($result);
+
+		return $notified_users;
+	}
+
 	/**
 	* Parse the queue and notify the users
 	*/
 	public function notify()
 	{
+		$insert_buffer = new \phpbb\db\sql_insert_buffer($this->db, $this->notification_emails_table);
+
+		/** @var \phpbb\notification\type\type_interface $notification */
+		foreach ($this->queue as $notification)
+		{
+			$data = $this->clean_data($notification->get_insert_array());
+			$insert_buffer->insert($data);
+		}
+
+		$insert_buffer->flush();
+
 		return $this->notify_using_messenger(NOTIFY_EMAIL);
 	}
+
+	/**
+	* {@inheritdoc}
+	*/
+	public function mark_notifications($notification_type_id, $item_id, $user_id, $time = false, $mark_read = true)
+	{
+		$sql = 'DELETE FROM ' . $this->notification_emails_table . '
+			WHERE ' . (($notification_type_id !== false) ? (is_array($notification_type_id) ? $this->db->sql_in_set('notification_type_id', $notification_type_id) : 'notification_type_id = ' . $notification_type_id) : '1=1') .
+			(($user_id !== false) ? ' AND ' . (is_array($user_id) ? $this->db->sql_in_set('user_id', $user_id) : 'user_id = ' . (int) $user_id) : '') .
+			(($item_id !== false) ? ' AND ' . (is_array($item_id) ? $this->db->sql_in_set('item_id', $item_id) : 'item_id = ' . (int) $item_id) : '');
+		$this->db->sql_query($sql);
+	}
+
+	/**
+	* {@inheritdoc}
+	*/
+	public function mark_notifications_by_parent($notification_type_id, $item_parent_id, $user_id, $time = false, $mark_read = true)
+	{
+		$sql = 'DELETE FROM ' . $this->notification_emails_table . '
+			WHERE ' . (($notification_type_id !== false) ? (is_array($notification_type_id) ? $this->db->sql_in_set('notification_type_id', $notification_type_id) : 'notification_type_id = ' . $notification_type_id) : '1=1') .
+			(($user_id !== false) ? ' AND ' . (is_array($user_id) ? $this->db->sql_in_set('user_id', $user_id) : 'user_id = ' . (int) $user_id) : '') .
+			(($item_parent_id !== false) ? ' AND ' . (is_array($item_parent_id) ? $this->db->sql_in_set('item_parent_id', $item_parent_id, false, true) : 'item_parent_id = ' . (int) $item_parent_id) : '');
+		$this->db->sql_query($sql);
+	}
+
+	/**
+	 * Clean data to contain only what we need for email notifications table
+	 *
+	 * @param array $data Notification data
+	 * @return array Cleaned notification data
+	 */
+	protected function clean_data(array $data)
```

Comment:
Any chance this could become a `static public` method? Looks like something that might be used outside the class.